exec: introduce two panic functions to use within the exec package

yuzefovich opened the pull request:
Previously, all errors (except for the ones originated below SQL
layer) without PG codes would be treated as "internal errors"
and would be returned to the client with the stack trace. However,
in a few places, those errors are not caused by the vectorized
engine, and we would like to simply propagate the errors.

In order to distinguish whether an error is caused by the vectorized
engine, two new "panic" functions are introduced. All code within
pkg/sql/exec is now required to use one of the two, and the usage
of Golang's panic function is prohibited with a linter.

Fixes: #39422.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/39434)
<!-- Reviewable:end -->

jordanlewis commented:
I think this will work, but it seems like it might be easy for the next person to forget to do this properly.

This is why I was thinking originally that it might be nice to have a couple of methods that we use within `exec` instead of raw panics. I think one could be for assertion errors and one could be for pass-through errors. Then, we could lint out all explicit `panics` and force people to decide which of the two to use every time they need to raise an exception.

yuzefovich commented:
Good point, I'll refactor the code.

yuzefovich commented:
Refactored, RFAL.

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39434)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/workload/dep_test.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39434#-LlmKSQG1e_cp2F3oIee:-LlmKSQG1e_cp2F3oIef:b-wq63u3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8bbbb7000bd57bcc819575afeb7fdf377a38ce3/pkg/workload/dep_test.go#L28)):*
> ```Go
> 		[]string{
> 			`github.com/cockroachdb/cockroach/pkg/sql/exec/coldata`,
> 			`github.com/cockroachdb/cockroach/pkg/sql/exec/execerror`,
> ```

cc @danhhz for this addition


<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39434)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/workload/dep_test.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39434#-LlmKSQG1e_cp2F3oIee:-LlmN-f62a5DEFYCx1cM:b-452t5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8bbbb7000bd57bcc819575afeb7fdf377a38ce3/pkg/workload/dep_test.go#L28)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

cc @danhhz for this addition
</blockquote></details>

huh, that package depends in turn on github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode and github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror but I don't see them in this list. I wonder how this is passing

anyway, if i understand what this package if for (error handling in our sql execution?) this doesn't really seem like an appropriate package for workload to depend on. i think we should find a way to avoid it


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39434)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @danhhz, and @jordanlewis)

___
*[pkg/workload/dep_test.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39434#-LlmKSQG1e_cp2F3oIee:-LlmQ_MS9haAIGWwZ_FY:bmm9jg3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8bbbb7000bd57bcc819575afeb7fdf377a38ce3/pkg/workload/dep_test.go#L28)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

huh, that package depends in turn on github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode and github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror but I don't see them in this list. I wonder how this is passing

anyway, if i understand what this package if for (error handling in our sql execution?) this doesn't really seem like an appropriate package for workload to depend on. i think we should find a way to avoid it
</blockquote></details>

Yeah, `deptest` didn't pass. I agree with you, and I'll refactor the code to not introduce new dependencies.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39434)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @danhhz, and @jordanlewis)

___
*[pkg/workload/dep_test.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39434#-LlmKSQG1e_cp2F3oIee:-Lln2b8g1O17FL62l_o1:b-4b6h1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8bbbb7000bd57bcc819575afeb7fdf377a38ce3/pkg/workload/dep_test.go#L28)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Yeah, `deptest` didn't pass. I agree with you, and I'll refactor the code to not introduce new dependencies.
</blockquote></details>

I refactored `exec/execerror` package so that it doesn't bring in `pgwire` packages. We want to enforce a particular panic convention onto everything that is within `sql/exec` (including `exec/execgen` and `exec/coldata` that are dependencies here). Dan, are you ok with adding only `exec/execerror` here?


<!-- Sent from Reviewable.io -->

danhhz commented:
coldata is currently a general-purpose columnar data format and is used like that by workload. it's not really correct for it to be concerned with sql execution, even though that is it's primary user and where it was extracted from.

i'm not going to stand in the way on progress on this, but i definitely prefer to find some other way to accomplish what you'd like. certainly you're not threading these wrappers down into every library that's transitively used by sql. how would you and @jordanlewis feel about us moving coldata (and colserde) out from under pkg/sql?

jordanlewis commented:
@yuzefovich, is it the case that the `coldata` package can never produce an expected exception? It seems like it to me - in what case would we want to present a pretty (non-stack-trace-y) error to the user from a panic thrown by `coldata`? For example, I only see "internal error" exceptions coming from `coldata` in this patch, which implies that replacing those with `panic` (and acquiescing to @danhhz's reasonable desire) would change no behavior.

yuzefovich commented:
Ok, I agree, we can have regular panics in `coldata` and skip it in the linter. What about `execgen`? Why is it a dependency of `workload`? I think that we want to monitor `execgen` with a linter.

jordanlewis commented:
I don't know - I think it's because we generate some code in workload right now? In any case, as long as we're not making it worse it's okay for now, right Dan? :)

yuzefovich commented:
Oh, it's a dependency because we have `vec_tmpl` in `coldata`.

danhhz commented:
oof, I would love to get rid of that execgen dependency but that's a much bigger change. do what you have to do for now and we'll clean that up later

danhhz commented:
another thought. execgen is not used at runtime by workload, this file is behind a build tag that we ignore. i wonder how hard it would be to thread knowledge of the "execgen_template" tag into the lint test and ignore those files

yuzefovich commented:
Ok, the build is green, so it is RFAL.

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 46 of 61 files at r1, 17 of 17 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39434#-:-LmAA_3t-YcVLL9nfmhK:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz, @jordanlewis, @rafiss, and @yuzefovich)

___
*[pkg/testutils/lint/lint_test.go, line 1462 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39434#-LmAA9z23JP8POD4gMq4:-LmAA9z23JP8POD4gMq5:b-36dvkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5fa77c3bebf48b2221b2bad43f8569f12ac76ea/pkg/testutils/lint/lint_test.go#L1462)):*
> ```Go
> 			"--",
> 			"sql/exec",
> 			":!sql/exec/coldata/",
> ```

Why are these files (minus `execerror`) excluded?. Maybe add a comment.
___
*[pkg/sql/distsqlrun/vectorized_panic_propagation_test.go, line 80 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39434#-LmA6h0CAl-DK0RXVgwE:-LmA6h0CAl-DK0RXVgwF:b7o65go) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5fa77c3bebf48b2221b2bad43f8569f12ac76ea/pkg/sql/distsqlrun/vectorized_panic_propagation_test.go#L80)):*
> ```Go
> 	}()
> 	if panicEmitted {
> 		t.Fatalf("VectorizedInternalPanic was not caught")
> ```

Suggestion: these tests could be more concise with `require.{Not}Panics()` if you choose.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39434)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @danhhz, @jordanlewis, and @rafiss)

___
*[pkg/testutils/lint/lint_test.go, line 1462 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39434#-LmAA9z23JP8POD4gMq4:-LmB7K2T1iuwV5CRcvOe:b-6kl47p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5fa77c3bebf48b2221b2bad43f8569f12ac76ea/pkg/testutils/lint/lint_test.go#L1462)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Why are these files (minus `execerror`) excluded?. Maybe add a comment.
</blockquote></details>

Good point, I left a comment.

The issue is that `pkg/workload` uses `exec/coldata` and `exec/types` packages, and we don't want to introduce a dependency on `execerror` into that. I think we might want to move those packages outside of `sql/exec` because of this, but that should be a separate PR.
___
*[pkg/sql/distsqlrun/vectorized_panic_propagation_test.go, line 80 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39434#-LmA6h0CAl-DK0RXVgwE:-LmB46oz8uhOdT0AW7tG:b-4pj5my) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5fa77c3bebf48b2221b2bad43f8569f12ac76ea/pkg/sql/distsqlrun/vectorized_panic_propagation_test.go#L80)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Suggestion: these tests could be more concise with `require.{Not}Panics()` if you choose.
</blockquote></details>

Nice, done.


<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39434)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @danhhz, @jordanlewis, and @rafiss)

___
*[pkg/testutils/lint/lint_test.go, line 1462 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39434#-LmAA9z23JP8POD4gMq4:-LmBCiQj9qcCZ5Rhq1P3:b-10y8m5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5fa77c3bebf48b2221b2bad43f8569f12ac76ea/pkg/testutils/lint/lint_test.go#L1462)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Good point, I left a comment.

The issue is that `pkg/workload` uses `exec/coldata` and `exec/types` packages, and we don't want to introduce a dependency on `execerror` into that. I think we might want to move those packages outside of `sql/exec` because of this, but that should be a separate PR.
</blockquote></details>

https://github.com/cockroachdb/cockroach/pull/39569


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39434)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @danhhz, @jordanlewis, and @rafiss)

___
*[pkg/testutils/lint/lint_test.go, line 1462 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39434#-LmAA9z23JP8POD4gMq4:-LmBj1c_0HbmRtKvRJrq:bivlafw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5fa77c3bebf48b2221b2bad43f8569f12ac76ea/pkg/testutils/lint/lint_test.go#L1462)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

https://github.com/cockroachdb/cockroach/pull/39569
</blockquote></details>

Nice! I'll wait for #39569 to go in before merging this.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
exec: introduce two panic functions to use within the exec package

Previously, all errors (except for the ones originated below SQL
layer) without PG codes would be treated as "internal errors"
and would be returned to the client with the stack trace. However,
in a few places, those errors are not caused by the vectorized
engine, and we would like to simply propagate the errors.

In order to distinguish whether an error is caused by the vectorized
engine, two new "panic" functions are introduced. All code within
pkg/sql/exec is now required to use one of the two, and the usage
of Golang's panic function is prohibited with a linter.

Release note: None

yuzefovich commented:
Thanks for the review/input!

bors r+
